src/fs/mod.rs

```diff
@@ -106,44 +118,70 @@ impl Filesystem {
         let chunk_size = CHUNK_SIZE; // file.block_size as usize;
         let chunk_index = offset / chunk_size;
 
-        if chunk_index >= file.blocks.len() || op.size() == 0 {
+        if chunk_index >= file_metadata.blocks.len() || op.size() == 0 {
             // reading after the end of the file
             let data: &[u8] = &[];
             return Ok(req.reply(data)?);
         }
 
         // offset inside the file
-        let mut offset = offset - (chunk_index * chunk_size);
+        let mut internal_offset = offset - (chunk_index * chunk_size);
         let mut cache = self.cache.clone();
         let mut buf: Vec<u8> = vec![0; size];
         let mut total = 0;
 
-        'blocks: for block in file.blocks.iter().skip(chunk_index) {
-            let (_, mut fd) = match cache.get(block).await {
-                Ok(out) => out,
-                Err(_) => {
-                    return Ok(req.reply_error(libc::EIO)?);
+        'blocks: for block in file_metadata.blocks.iter().skip(chunk_index) {
+            // fhash works as a key inside the LRU
+            let fhash = block.hash.clone();
+
+            // getting the file descriptor from the LRU or from the cache if not found in the LRU
+            let lrufd = self.lruf.lock().await.pop(&fhash);
+
+            let (mut fd, block_size) = match lrufd {
+                Some((descriptor, bsize)) => (descriptor, bsize),
+                None => {
+                    let (bsize, descriptor) = match cache.get(block).await {
+                        Ok(out) => out,
+                        Err(_) => {
+                            return Ok(req.reply_error(libc::EIO)?);
+                        }
+                    };
+                    (descriptor, bsize)
                 }
             };
-            fd.seek(SeekFrom::Start(offset as u64)).await?;
+
+            // seek to the position <offset>
+            fd.seek(SeekFrom::Start(internal_offset as u64)).await?;
 
             loop {
+                // read the file bytes into buf
                 let read = match fd.read(&mut buf[total..]).await {
                     Ok(n) => n,
                     Err(_) => {
                         return Ok(req.reply_error(libc::EIO)?);
                     }
                 };
+
+                // calculate the total size and break if the required bytes (=size) downloaded
                 total += read;
                 if total >= size {
+                    let read_size = fd.stream_position().await?;
```

Comment:
It's not a good idea to use `stream_position` because it's another `systemall` (plus you await) while u can perfectly track the current stream position of this chunk because you do `seek` (line 54) and then you know the  (exact sizes of the reads) you did on the file so the current offset is always (seek + sum(all reads on the file))

I am 100% sure i explained this before either in this PR or in one of the zoom calls. And you still did it differently.